rollup/circuitcapacitychecker/impl.go

```diff
@@ -65,13 +67,14 @@ func (ccc *CircuitCapacityChecker) ApplyTransaction(traces *types.BlockTrace) (*
 		return nil, ErrUnknown
 	}
 
-	tracesByt, err := json.Marshal(traces)
+	ccc.jsonBuffer.Reset()
+	err := json.NewEncoder(&ccc.jsonBuffer).Encode(traces)
```

Comment:
does it cause race conditions between `ApplyTransaction` and `ApplyBlock` ?

Reply:
Nope, both functions require the ccc lock to be held.

Reply:
o, ic.